src/smartpeak/include/SmartPeak/ui/SessionFilesWidget.h

```diff
@@ -0,0 +1,106 @@
+// --------------------------------------------------------------------------
+//   SmartPeak -- Fast and Accurate CE-, GC- and LC-MS(/MS) Data Processing
+// --------------------------------------------------------------------------
+// Copyright The SmartPeak Team -- Novo Nordisk Foundation 
+// Center for Biosustainability, Technical University of Denmark 2018-2021.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Douglas McCloskey, Bertrand Boudaud $
+// $Authors: Douglas McCloskey, Bertrand Boudaud $
+// --------------------------------------------------------------------------
+#pragma once
+
+#include <SmartPeak/ui/Widget.h>
+#include <SmartPeak/core/ApplicationHandler.h>
+#include <SmartPeak/ui/FilePicker.h>
+#include <misc/cpp/imgui_stdlib.h>
+#include <string>
+#include <vector>
+
+namespace SmartPeak
+{
+  class SessionFilesWidget : public Widget
+  {
+  public:
+
+    enum class Mode
+    {
+      ECreation,
+      EModification
+    };
+
+    SessionFilesWidget(ApplicationHandler& application_handler, SessionFilesWidget::Mode mode);
+
+    void draw() override;
+    void open(const Filenames& filenames, const std::set<std::string>& requirements = {});
+
+  protected:
+
+    struct FileEditorFields
+    {
+      std::string text_editor_hint_;
+      std::string text_editor_;
+      std::string text_editor_last_update_; // used to detect a change in the textfield (cb does not work properly)
+      std::string original_filepath_;
+      bool embedded_ = false;
+      bool original_embedded_ = false;
+      bool file_exists_ = false;
+      std::string combo_file_mode_ = "";
+    };
+
+    Filenames filenames_;
+    std::optional<std::string> popup_file_picker_;
```

Comment:
No it's not due to the SQL light integration code, I just find it useful. std::optional is used quite a lot when something may not be available, or when a method can return an object or nothing (in case of error).
in that case, a flag was needed to say I want to open the filepicker with that file_id (ImGui cannot open correctly the filepicker when it's needed so we need a variable to postpone it). if the option is not set, setting the option on will open the filepicker with the right file_id. if the option is not set, the file picker will not open. So in the end, it avoids the declaration of 2 variables for one action.